modules/biobank/php/containerdao.class.inc

```diff
@@ -0,0 +1,534 @@
+                     bcp.Coordinate,
+                     bc.DateTimeUpdate,
+                     bc.DateTimeCreate
+                  FROM biobank_container bc
+                  LEFT JOIN biobank_container_parent bcp
+                    USING (ContainerID)
+                  JOIN biobank_container_type bct
+                    USING (ContainerTypeID)
+                  WHERE bc.ContainerID=:i';
+        $container = $this->db->pselectrow($query, array('i' => $id));
+
+        return $container;
+    }
+
+    /**
+     * Returns an array of all the Child Container IDs associated with
+     * the given Container ID from the biobank_container_parent table.
+     *
+     * @param int $id of Container
+     *
+     * @return array $childContainerIds List of Container IDs that are
+     *                                  children of the Container ID
+     */
+    private function getChildContainerIds(int $id) : array
+    {
+         $query       = 'SELECT ContainerID
+                   FROM biobank_container_parent
+                   WHERE ParentContainerID=:i';
+         $childContainerIds = $this->db->pselectCol($query, array('i' => $id));
+
+         return $childContainerIds;
+    }
+
+    // XXX: Currently this function is never used with any conditions passed as
+    // parameters. Decide if this is a necessary feature. It is likely useful,
+    // but I need to figure out how to integrate it wit the dataprovisioner.
+    /**
+     * This will select all containers from the database that match the
+     * attribute values passed by $conditions and will return an array
+     * of container objects.
+     *
+     * @param array $conditions Conditions to be met by the query
+     *
+     * @return array $containers List of Container Objects that match the query
+     */
+    public function selectInstances(array $conditions = null) : array
+    {
+        $query  = 'SELECT ContainerID
+                  FROM biobank_container
+                  LEFT JOIN biobank_container_parent
+                    USING (ContainerID)
+                  JOIN biobank_container_type
+                    USING (ContainerTypeID)';
+        $query .= $conditions ? ' WHERE '.$this->db->_implodeWithKeys(' AND ', $conditions) : '';
+
+        $result     = $this->db->pselect($query, array());
+        $containers = array();
+        if (!empty($result)) {
+            foreach ($result as $row) {
+                $id        = intval($row['ContainerID']);
+                $container = $this->getInstanceFromId($id);
+                $containers[$id] = $container;
+            }
+        }
+
+        return $containers;
+    }
+
+    /**
+     * Instantiates an ArrayIterator class that is composed of all the Container
+     * Objects.
+     *
+     * @return traversable Iterator of Container Objects
+     */
+    protected function getAllInstances() : \Traversable
+    {
+        return new \ArrayIterator($this->selectInstances());
```

Comment:
Are there instances where you need `selectInstances` is used for anything other than foreach? If not, it might be easier to directly return a `\Traversable` from there rather than returning an array that needs to get converted into an iterable with `\ArrayIterator`

Reply:
So are the end of `selectInstances()` I would `return new\ArrayIterator($containers)`
and then at the end of `getAllInstances()` I would `return `$this->selectInstances()`?

Reply:
@driusan I'm also curious to understand what you mean here

Reply:
You can use 2 things as the target of a foreach loop: `array`, or `\Traversable` objects (since it's stupid that `array`'s aren't considered `\Traversable` (because they aren't objects), PHP also later added `iterable` as a typehint which means "array or Traversable"...) An `\ArrayIterator` is something that takes an array and converts it to an `\Iterator` (an iterator being a type of Traversable.), but the entire array still needs to be held in memory for it to work.

If the only reason to call selectInstances is to later foreach over them, then you don't need to load things into an array in the first place (which requires storing the entire data set in memory). PDOStatements are already `\Traversable` objects.